File: basicAlgorithms/_tortoisehare.js

/*
Here we use the tortoise and the hare approach to find the middle point.

Basically the tortoise and hare consist on one cursor moving at two nodes pace
and the other at one node pace.

In this example we are using it to find the middle point.
*/

function ListNode(val,next = null){
    this.val = (val===undefined ? 0 : val);
    this.next = (next===undefined ? null : next);
}

function middle(head){
    let liebre = head;
    let tortuga = head;

    while(liebre != null && liebre.next != null){
        liebre = liebre.next.next;
        tortuga = tortuga.next;
    }

    return tortuga;
}

File: basicAlgorithms/_tortoisehare.cpp
#include <iostream>

struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* middlePoint(ListNode* head){
    ListNode* liebre = head;
    ListNode* tortuga = head;

    while(liebre != nullptr && liebre->next != nullptr){
        liebre = liebre->next->next;
        tortuga = tortuga->next;
    }
    return tortuga;
}